File: src/pages/student/universityModel.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import { markAmount } from '../../data/assessmentScheme.ts'
import type { ModuleScheduleStore } from '../../data/moduleSchedule.ts'
import type { ModuleSubjectStore } from '../../data/moduleSubjects.ts'
import type { University } from '../../data/universities.ts'
import {
  buildDemoStudentUniversityProjection,
  normalizeStudentUniversityProjection,
  type StudentUniversityProjection,
} from './universityModel.ts'

const now = new Date('2026-08-26T12:00:00')

test('live projection normalizes only the authenticated cohort and keeps unavailable totals honest', () => {
  const projection: StudentUniversityProjection = {
    profile: { studentId: 's1', universityId: 'hu', year: 'Year 2', yearId: 'HU_Y2', group: 'A' },
    university: { id: 'hu', name: 'Helwan University', short: 'HU', region: 'Cairo' },
    year: { id: 'HU_Y2', year: 'Year 2', active: true, terms: ['Term 1'] },
    status: 'ready',
    terms: [{
      term: 'Term 1',
      modules: [
        {
          id: 'inh',
          name: 'INH 201',
          moduleId: 'INH 201',
          term: 'Term 1',
          labels: ['verified'],
          assessment: {
            status: 'exact',
            declaredTotal: 360,
            componentTotal: 360,
            total: 360,
            displayTotal: 360,
            credits: 6,
            components: [{ id: 'written', label: 'Final written', kind: 'final-written', marks: 347.5 }],
          },
          subjects: [{ id: 's1', name: 'Anatomy', labels: ['verified'], coverage: { articleIds: ['ART-1'] }, children: [{ id: 's1-a', name: 'Neuroanatomy', children: [] }] }],
          schedule: [{ id: 'row-1', type: 'lecture', title: 'Safe row', date: '2026-12-15', startTime: '09:00', endTime: '10:00', labels: ['inferred', 'carried-forward'], links: { articleIds: ['ART-2'], conceptIds: ['CON-2'] } }],
          coverage: { counts: { articleIds: 2, conceptIds: 1 } },
        },
        {
          id: 'crs',
          name: 'CRS 204',
          moduleId: 'CRS 204',
          term: 'Term 1',
          assessment: { status: 'partial', displayTotal: 'unavailable', components: [] },
          subjects: [],
          schedule: [],
          coverage: {},
        },
      ],
    }],
    modules: [
      { id: 'inh', name: 'INH 201', moduleId: 'INH 201', term: 'Term 1' },
      { id: 'crs', name: 'CRS 204', moduleId: 'CRS 204', term: 'Term 1' },
    ],
  }

  const map = normalizeStudentUniversityProjection(projection, now)
  assert.equal(map.university?.id, 'hu')
  assert.equal(map.year?.id, 'HU_Y2')
  assert.equal(map.modules.length, 2)
  assert.equal(map.modules[0].assessment.displayTotal, '360')
  assert.equal(map.modules[0].assessment.components[0].displayMarks, '347.5')
  assert.equal(map.modules[0].schedule[0].linkCount, 2)
  assert.deepEqual(map.modules[0].badges, ['verified', 'carried-forward', 'inferred'])
  assert.equal(map.modules[1].assessment.displayTotal, 'unavailable')
  assert.deepEqual(map.modules[1].badges, ['needs-marks', 'needs-schedule'])
  assert.equal(map.totals.marks, 360)
  assert.equal(map.totals.marksUnavailable, true)
})

test('being-verified projection stays empty without inventing a catalogue year', () => {
  const map = normalizeStudentUniversityProjection({
    profile: { studentId: 's1', universityId: 'hu', year: 'Year 4', yearId: null, group: null },
    university: { id: 'hu', name: 'Helwan University', short: 'HU', region: 'Cairo' },
    year: null,
    terms: [],
    modules: [],
    status: 'being_verified',
  }, now)
  assert.equal(map.status, 'being_verified')
  assert.equal(map.year, null)
  assert.deepEqual(map.modules, [])
})

test('demo projection uses local stores but preserves flexible schemes and schedule labels', () => {
  const university: University = {
    id: 'hu',
    name: 'Helwan University',
    short: 'HU',
    region: 'Cairo',
    years: [{
      id: 'HU_Y2',
      year: 'Year 2',
      students: 0,
      terms: ['Term 1'],
      courses: [
        { id: 'inh', name: 'INH 201', block: 'Term 1', moduleId: 'INH 201', term: 'Term 1' },
        { id: 'crs', name: 'CRS 204', block: 'Term 1', moduleId: 'CRS 204', term: 'Term 1' },
      ],
    }],
  }
  const subjects: ModuleSubjectStore = {
    'hu:HU_Y2:inh': [{
      id: 'micro',
      name: 'Microbiology',
      marks: { writtenEndOfModule: 31, writtenEndOfYear: 15, practicalEndOfModule: 0, practicalEndOfYear: 0 },
      curriculum: { articleIds: ['ART-1'], questionIds: [], practicalIds: [], topicNodeIds: [], conceptIds: [], resourceIds: [] },
    }],
  }
  const schedules: ModuleScheduleStore = {
    'hu:HU_Y2:inh': [{
      id: 'old',
      type: 'lecture',
      title: 'Intro',
      date: '2026-01-01',
      startTime: '09:00',
      endTime: '10:00',
      location: '',
      moduleNumber: '',
      topicIds: ['ART-2'],
      notes: '',
      automaticQuestions: true,
      automaticPracticals: true,
      automaticQuestionIds: [],
      automaticPracticalIds: [],
      manualQuestionIds: [],
      manualPracticalIds: [],
      completed: false,
      carryForward: { sourceDate: '2025-01-01', sourceCycle: '2025', carriedForwardFrom: 'old-2025', targetCycle: '2026' },
      provenance: { evidenceState: 'inferred', sourceRefs: [] },
    }],
  }
  const projection = buildDemoStudentUniversityProjection(university, university.years[0], subjects, schedules, {
    'hu:HU_Y2:inh': {
      declaredTotal: markAmount(100) ?? undefined,
      reconciliationStatus: 'exact',
      components: [{ id: 'eom', label: 'EOM', kind: 'eom', marks: markAmount(100) ?? 0 }],
    },
  })
  const map = normalizeStudentUniversityProjection(projection, now)
  assert.equal(map.modules[0].assessment.displayTotal, '100')
  assert.equal(map.modules[0].schedule[0].labels.includes('carried-forward'), true)
  assert.equal(map.modules[0].schedule[0].labels.includes('inferred'), true)
  assert.equal(map.modules[0].schedule[0].linkCount, 1)
  assert.equal(map.modules[1].assessment.displayTotal, 'unavailable')
})

File: android/app/src/test/java/com/synapse/app/core/university/UniversityModelTest.kt
package com.synapse.app.core.university

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant

/**
 * Ported test vectors from `universityModel.test.ts`'s first two cases (the
 * live-projection path). The third vector (`buildDemoStudentUniversityProjection`)
 * is intentionally not ported — see the doc comment on [UniversityModel.kt]
 * (`com.synapse.app.core.university`) for why.
 */
class UniversityModelTest {

    private val now = Instant.parse("2026-08-26T12:00:00Z")

    @Test
    fun liveProjectionNormalizesOnlyTheAuthenticatedCohortAndKeepsUnavailableTotalsHonest() {
        val projection = StudentUniversityProjection(
            profile = ProjectionProfile(studentId = "s1", universityId = "hu", year = "Year 2", yearId = "HU_Y2", group = "A"),
            university = ProjectionUniversity(id = "hu", name = "Helwan University", short = "HU", region = "Cairo"),
            year = ProjectionYear(id = "HU_Y2", year = "Year 2", active = true, terms = listOf("Term 1")),
            status = "ready",
            terms = listOf(
                ProjectionTerm(
                    term = "Term 1",
                    modules = listOf(
                        ProjectionModule(
                            id = "inh",
                            name = "INH 201",
                            moduleId = "INH 201",
                            term = "Term 1",
                            labels = listOf("verified"),
                            assessment = ProjectionAssessment(
                                status = "exact",
                                declaredTotal = 360.0,
                                componentTotal = 360.0,
                                total = 360.0,
                                credits = 6.0,
                                components = listOf(
                                    ProjectionAssessmentComponent(id = "written", label = "Final written", kind = "final-written", marks = 347.5),
                                ),
                            ),
                            subjects = listOf(
                                ProjectionSubject(
                                    id = "s1",
                                    name = "Anatomy",
                                    labels = listOf("verified"),
                                    coverage = ProjectionCoverage(articleIds = listOf("ART-1")),
                                    children = listOf(ProjectionSubject(id = "s1-a", name = "Neuroanatomy")),
                                ),
                            ),
                            schedule = listOf(
                                ProjectionScheduleRow(
                                    id = "row-1",
                                    type = "lecture",
                                    title = "Safe row",
                                    date = "2026-12-15",
                                    startTime = "09:00",
                                    endTime = "10:00",
                                    labels = listOf("inferred", "carried-forward"),
                                    links = ProjectionScheduleLinks(articleIds = listOf("ART-2"), conceptIds = listOf("CON-2")),
                                ),
                            ),
                            coverage = ProjectionCoverage(counts = mapOf("articleIds" to 2, "conceptIds" to 1)),
                        ),
                        ProjectionModule(
                            id = "crs",
                            name = "CRS 204",
                            moduleId = "CRS 204",
                            term = "Term 1",
                            assessment = ProjectionAssessment(status = "partial", components = emptyList()),
                        ),
                    ),
                ),
            ),
            modules = listOf(
                ProjectionModuleSummary(id = "inh", name = "INH 201", moduleId = "INH 201", term = "Term 1"),
                ProjectionModuleSummary(id = "crs", name = "CRS 204", moduleId = "CRS 204", term = "Term 1"),
            ),
        )

        val map = normalizeStudentUniversityProjection(projection, now)

        assertEquals("hu", map.university?.id)
        assertEquals("HU_Y2", map.year?.id)
        assertEquals(2, map.modules.size)
        assertEquals("360", map.modules[0].assessment.displayTotal)
        assertEquals("347.5", map.modules[0].assessment.components[0].displayMarks)
        assertEquals(2, map.modules[0].schedule[0].linkCount)
        assertEquals(listOf(ModuleBadge.Verified, ModuleBadge.CarriedForward, ModuleBadge.Inferred), map.modules[0].badges)
        assertEquals("unavailable", map.modules[1].assessment.displayTotal)
        assertEquals(listOf(ModuleBadge.NeedsMarks, ModuleBadge.NeedsSchedule), map.modules[1].badges)
        assertEquals(360.0, map.totals.marks)
        assertTrue(map.totals.marksUnavailable)
    }

    @Test
    fun beingVerifiedProjectionStaysEmptyWithoutInventingACatalogueYear() {
        val projection = StudentUniversityProjection(
            profile = ProjectionProfile(studentId = "s1", universityId = "hu", year = "Year 4", yearId = null, group = null),
            university = ProjectionUniversity(id = "hu", name = "Helwan University", short = "HU", region = "Cairo"),
            year = null,
            terms = emptyList(),
            modules = emptyList(),
            status = "being_verified",
        )

        val map = normalizeStudentUniversityProjection(projection, now)

        assertEquals("being_verified", map.status)
        assertEquals(null, map.year)
        assertTrue(map.modules.isEmpty())
    }
}
